File: 0205-isomorphic-strings/two-way-mapping/python/solution.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        map_s_to_t, map_t_to_s = {}, {}

        for char1, char2 in zip(s, t):
            if char1 in map_s_to_t and map_s_to_t[char1] != char2:
                return False
            if char2 in map_t_to_s and map_t_to_s[char2] != char1:
                return False
            map_s_to_t[char1] = char2
            map_t_to_s[char2] = char1
            
        return True

File: 0205-isomorphic-strings/two-way-mapping/cpp/solution.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.length() != t.length()) 
            return false;

        unordered_map<char, char> s_to_t_map;
        unordered_map<char, char> t_to_s_map;

        for (int idx = 0; idx < s.length(); idx++) {
            char char1 = s[idx];
            char char2 = t[idx];

            if (s_to_t_map.count(char1) && s_to_t_map[char1] != char2) 
                return false;
            
            if (t_to_s_map.count(char2) && t_to_s_map[char2] != char1)
                return false;

            s_to_t_map[char1] = char2;
            t_to_s_map[char2] = char1;
        }
        return true;
    }
};
